src/main/java/com/coveros/selenified/application/App.java

```diff
@@ -101,31 +100,31 @@
      * @throws MalformedURLException   If the provided hub address isn't a URL, this exception will
      *                                 be thrown
      */
-    public App(Browser browser, DesiredCapabilities capabilities,
+    public App(Capabilities capabilities,
                OutputFile file) throws InvalidBrowserException, MalformedURLException {
-        if (browser == null) {
-            this.browser = new Browser(BrowserName.NONE);
+        if (capabilities == null || capabilities.getBrowser() == null) {
+            this.browser = new Browser("None");
         } else {
-            this.browser = browser;
+            this.browser = capabilities.getBrowser();
         }
 
-        if (capabilities == null) {
-            this.capabilities = new DesiredCapabilities();
+        if (capabilities == null || capabilities.getDesiredCapabilities() == null) {
+            this.desiredCapabilities = new DesiredCapabilities();
         } else {
-            this.capabilities = capabilities;
+            this.desiredCapabilities = capabilities.getDesiredCapabilities();
         }
         this.file = file;
 
         // if we want to test remotely
         if (System.getProperty("hub") != null) {
-            driver = new RemoteWebDriver(new URL(System.getProperty("hub") + "/wd/hub"), this.capabilities);
+            driver = new RemoteWebDriver(new URL(System.getProperty("hub") + "/wd/hub"), this.desiredCapabilities);
         } else {
             try {
-                this.capabilities.setJavascriptEnabled(true);
+                this.desiredCapabilities.setJavascriptEnabled(true);
             } catch (NullPointerException e) {
                 log.error(e);
             }
-            driver = TestSetup.setupDriver(this.browser, this.capabilities);
+            driver = capabilities.setupDriver();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "capabilities" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/sonar/coding_rules#rule_key=squid%3AS2259)